### PR TITLE
compression benchmark

### PR DESCRIPTION
**Pull request by suiguoxin**
Filter pruning experiment with : **SimulatedAnnealing, NetAdapt, AutoCompress, L1Filter, L2Filter, FPGMPruner** on **cifar10** with **resnet18, resnet50 and vgg16**.

This #PR includes the following contents:
- experiment result presentation & analysis
- source code and instruction fore re-implementation
- experiment result in json format

For `ActivationAPoZRankFilterPruner`, `ActivationMeanRankFilterPruner` & `AGPPruner`, I plan to add them in this benchmark after refactoring.

### CODE REVIEW

**Comment by colorjam** [comment]
suggest to remove .json files

File: docs/en_US/Compressor/Benchmark.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+To provide an initial insight into the performance of various channel pruning algorithms,
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Benchmark.md:1]: Suggest to use term `channel pruning` and `filter pruning` to represent different pruning methods:
channel pruning:  prune input channel
filter pruning: prune output channel
Our current pruner implemented as filter pruning,   we may support channel pruning in the future.

File: examples/model_compress/auto_pruners_torch.py [hunk 2]
```diff
@@ -263,129 +298,140 @@ def evaluator(model):
...
-        m_speedup.speedup_model()
-        evaluation_result = evaluator(model)
-        print('Evaluation result (speed up model): %s' % evaluation_result)
-        result['speedup'] = evaluation_result
-
-        torch.save(model.state_dict(), os.path.join(args.experiment_data_dir, 'model_speed_up.pth'))
-        print('Speed up model saved to %s', args.experiment_data_dir)
-
-    with open(os.path.join(args.experiment_data_dir, 'performance.json'), 'w+') as f:
+    with open(os.path.join(args.experiment_data_dir, 'result.json'), 'w+') as f:
```
Comment (chicm-ms) [inline examples/model_compress/auto_pruners_torch.py:371]: Can we save all results into one json?  maybe use a key to identify different experiment?

File: examples/model_compress/models/cifar10/resnet.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import torch
+import torch.nn as nn
```
Comment (chicm-ms) [inline examples/model_compress/models/cifar10/resnet.py:2]: what is the difference between this file and torchvision.models.resnet ?

File: docs/en_US/Compressor/Benchmark.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+![](../../../examples/model_compress/experiment_result/img/performance_comparison_vgg16.png)
+
+CIFAR-10, ResNet18:
+
+![](../../../examples/model_compress/experiment_result/img/performance_comparison_resnet18.png)
+
+CIFAR-10, ResNet50:
+
+![](../../../examples/model_compress/experiment_result/img/performance_comparison_resnet50.png)
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Benchmark.md:43]: Is it easier to read if the X axis is changed to sparsity/flops ratio ?

File: examples/model_compress/auto_pruners_torch.py [hunk 5]
```diff
@@ -127,65 +130,89 @@ def test(model, device, criterion, val_loader):
...
+def get_trained_model_optimizer(args, device, train_loader, val_loader, criterion):
     if args.model == 'LeNet':
         model = LeNet().to(device)
-        optimizer = torch.optim.Adadelta(model.parameters(), lr=1)
-        scheduler = StepLR(optimizer, step_size=1, gamma=0.7)
-        for epoch in range(args.pretrain_epochs):
-            train(args, model, device, train_loader,
-                  criterion, optimizer, epoch)
-            scheduler.step()
+        if args.load_pretrained_model:
```
Comment (chicm-ms) [inline examples/model_compress/auto_pruners_torch.py:136]: Do we have the LeNet benchmark result?

File: examples/model_compress/auto_pruners_torch.py [hunk 6]
```diff
@@ -127,65 +130,89 @@ def test(model, device, criterion, val_loader):
...
-            [args.test_batch_size, 3, 32, 32]).to(device)
-
+        dummy_input = torch.randn([args.test_batch_size, 3, 32, 32]).to(device)
     return dummy_input
 
 
+def get_input_size(dataset):
+    if dataset == 'mnist':
+        input_size = (1, 1, 28, 28)
+    elif dataset in ['cifar10', 'imagenet']:
```
Comment (chicm-ms) [inline examples/model_compress/auto_pruners_torch.py:205]: why input size for imagenet is 32x32?  do we resize the image for imagenet?

File: docs/en_US/Compressor/Benchmark.md (hunk 1 above)
Comment (QuanluZhang) [inline docs/en_US/Compressor/Benchmark.md:1]: better to add title for this doc, for example "Comparison of Pruning Algorithms"

File: docs/en_US/Compressor/Benchmark.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+* Datasets: CIFAR-10
+
+* Pruners: 
+    - These pruners are included:
+        - Pruners with scheduling : SimulatedAnnealing, NetAdapt, AutoCompress
+        - One-shot pruners: L1Filter, L2Filter, FPGMPruner
+    - Only **channel pruning** performances are compared here. 
+
+    For the auto-pruners, `L1FilterPruner` is used as the base algorithm. That is to say, after the sparsities distribution among the layers is decided by the scheduling algorithm, `L1FilterPruner` is used to performn real pruning.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Benchmark.md:20]: it is not clear what are the auto-pruners? Pruners with scheduling?

File: docs/en_US/Compressor/Benchmark.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+The experiments are performed with the following pruners/datasets/models:
+
+* Models: [VGG16, ResNet18, ResNet50](https://github.com/microsoft/nni/tree/master/examples/model_compress/models/cifar10)
+
+* Datasets: CIFAR-10
+
+* Pruners: 
+    - These pruners are included:
+        - Pruners with scheduling : SimulatedAnnealing, NetAdapt, AutoCompress
+        - One-shot pruners: L1Filter, L2Filter, FPGMPruner
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Benchmark.md:17]: better to mention how to set each layer's sparsity for these one-shot pruners

File: docs/en_US/Compressor/Benchmark.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+For each dataset/model/pruner combination, we prune the model to different levels by setting a series of target sparsities for the pruner. 
+
+Here we plot both **Number of Weights - Performances** curve and **FLOPs - Performance** curve. 
+As a reference, we also plot the result declared in the paper [AutoCompress: An Automatic DNN Structured Pruning Framework for Ultra-High Compression Rates](http://arxiv.org/abs/1907.03141) for models VGG16 and ResNet18 on CIFAR-10.
+
+The experiment result are shown in the following figures:
+
+CIFAR-10, VGG16:
+
+![](../../../examples/model_compress/experiment_result/img/performance_comparison_vgg16.png)
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Benchmark.md:35]: better to use different node types (not just different colors) for each pruner.

File: docs/en_US/Compressor/Benchmark.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+CIFAR-10, ResNet50:
+
+![](../../../examples/model_compress/experiment_result/img/performance_comparison_resnet50.png)
+
+## Analysis
+
+From the experiment result, we get the following conclusions:
+
+* Given the constraint on the number of parameters, the pruners with scheduling ( `AutoCompress` , `SimualatedAnnealing` ) performs better than the others when the constraint is strict. However, they have no such advantage in FLOPs/Performances comparison since only number of parameters constraint is considered in the optimization process; 
+* The basic algorithms `L1FilterPruner` , `L2FilterPruner` , `FPGMPruner` performs very similarly in these experiments;
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Benchmark.md:50]: good summary. for this one, can I say for some model/dataset one-shot pruners perform similar to the pruners with scheduling, even though they set the same sparsity for every layer?

File: docs/en_US/Compressor/Benchmark.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+* Given the constraint on the number of parameters, the pruners with scheduling ( `AutoCompress` , `SimualatedAnnealing` ) performs better than the others when the constraint is strict. However, they have no such advantage in FLOPs/Performances comparison since only number of parameters constraint is considered in the optimization process; 
+* The basic algorithms `L1FilterPruner` , `L2FilterPruner` , `FPGMPruner` performs very similarly in these experiments; 
+* `NetAdaptPruner` can not achieve very high compression rate. This is caused by its mechanism that it prunes only one layer each pruning iteration. This leads to un-acceptable complexity if the sparsity per iteration is much lower than the overall sparisity constraint.
+
+## Experiments Reproduction
+
+### Implementation Details
+
+* The experiment results are all collected with the default configuration of the pruners in nni.
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Benchmark.md:57]: what do you mean by 'default configuration'?

File: docs/en_US/model_compression.rst [hunk 12]
```diff
@@ -22,5 +22,6 @@ For details, please refer to the following tutorials:
     Automatic Model Compression <Compressor/AutoCompression>
     Model Speedup <Compressor/ModelSpeedup>
     Compression Utilities <Compressor/CompressionUtils>
+    Compression Benchmark <Compressor/Benchmark>
```
Comment (QuanluZhang) [inline docs/en_US/model_compression.rst:25]: this is not benchmark, it is benchmark result. So better to move it to `Use Cases and Solutions` under `Performance measurement, comparison and analysis`

File: examples/model_compress/experiment_result/analyze.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
+import argparse
```
Comment (QuanluZhang) [inline examples/model_compress/experiment_result/analyze.py:1]: the folder name `experiment_result` can be changed to `comparison_of_pruners`

File: examples/model_compress/auto_pruners_torch.py (hunk 5 above)
Comment (suiguoxin) [reply examples/model_compress/auto_pruners_torch.py:136]: No

File: examples/model_compress/auto_pruners_torch.py (hunk 6 above)
Comment (suiguoxin) [reply examples/model_compress/auto_pruners_torch.py:205]: There is an error here, changed to 256*256. Imagenet experiment is not performed in this PR.

File: docs/en_US/Compressor/Benchmark.md (hunk 4 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:43]: I don't see big difference...

File: examples/model_compress/models/cifar10/resnet.py (hunk 3 above)
Comment (suiguoxin) [reply examples/model_compress/models/cifar10/resnet.py:2]: > what is the difference between this file and torchvision.models.resnet ?
The only difference is here: https://github.com/suiguoxin/nni/blob/benchmark/examples/model_compress/models/cifar10/resnet.py#L69-L70 
We can not use torchvision.models.resnet because these args are not configurable.

File: docs/en_US/Compressor/Benchmark.md (hunk 1 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:1]: > Suggest to use term `channel pruning` and `filter pruning` to represent different pruning methods:
> channel pruning: prune input channel
> filter pruning: prune output channel
> Our current pruner implemented as filter pruning, we may support channel pruning in the future.

Thx, fixed.

File: docs/en_US/Compressor/Benchmark.md (hunk 7 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:20]: > it is not clear what are the auto-pruners? Pruners with scheduling?

Yes, fixed

File: docs/en_US/Compressor/Benchmark.md (hunk 1 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:1]: > better to add title for this doc, for example "Comparison of Pruning Algorithms"
Title 'Comparison of Filter Pruning Algorithms' added

File: examples/model_compress/auto_pruners_torch.py (hunk 2 above)
Comment (suiguoxin) [reply examples/model_compress/auto_pruners_torch.py:373]: > Can we save all results into one json? maybe use a key to identify different experiment?

Merged to 3 files. One file by every dataset/model combination.

**Comment by suiguoxin** [comment]
> are .json files necessary? suggest merge them into one file

Merged to 3 files. One file by every dataset/model combination.

**Comment by suiguoxin** [comment]
> are .json files necessary? suggest merge them into one file

Merged to 3 files. One file by every dataset/model combination.

File: examples/model_compress/experiment_result/analyze.py (hunk 13 above)
Comment (suiguoxin) [reply examples/model_compress/experiment_result/analyze.py:1]: > the folder name `experiment_result` can be changed to `comparison_of_pruners`

ok, thx

File: docs/en_US/Compressor/Benchmark.md (hunk 8 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:19]: > better to mention how to set each layer's sparsity for these one-shot pruners

Added here.

File: docs/en_US/Compressor/Benchmark.md (hunk 9 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md]: > better to use different node types (not just different colors) for each pruner.

Fixed

File: docs/en_US/Compressor/Benchmark.md (hunk 10 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:52]: > good summary. for this one, can I say for some model/dataset one-shot pruners perform similar to the pruners with scheduling, even though they set the same sparsity for every layer?

We cannot say pruners with scheduling and basic pruners are `similar` generally because they are different given different evaluation metrics.

File: docs/en_US/Compressor/Benchmark.md (hunk 11 above)
Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md:59]: > what do you mean by 'default configuration'?

Explained.

File: docs/en_US/model_compression.rst (hunk 12 above)
Comment (suiguoxin) [reply docs/en_US/model_compression.rst:25]: > this is not benchmark, it is benchmark result. So better to move it to `Use Cases and Solutions` under `Performance measurement, comparison and analysis`

Exactly, thanks

File: docs/en_US/Compressor/Benchmark.md (hunk 10 above)
Comment (tu1258) [reply docs/en_US/Compressor/Benchmark.md]: Isn't FLOPs related to just filter number and channel number(which are actually the same thing)? And filter number also represents parameters. How does it not improve FLOPs? I can imagine some architecture like xception won't improve because filter number reduction may not affect channel number. But it doesn't make sense for VGG16. Is it the same situation? Please correct me if I am wrong 😄

Comment (suiguoxin) [reply docs/en_US/Compressor/Benchmark.md]: > Isn't FLOPs related to just filter number and channel number(which are actually the same thing)? And filter number also represents parameters. How does it not improve FLOPs? I can imagine some architecture like xception won't improve because filter number reduction may not affect channel number. But it doesn't make sense for VGG16. Is it the same situation? Please correct me if I am wrong 😄

Thanks for your reply. FLOPs is also related to image resolution, the input size of different layers are different.

Comment (tu1258) [reply docs/en_US/Compressor/Benchmark.md]: Oh thanks for reminding me. Aren't Input size/feature map size almost the same size no matter which pruner? If so, the FLOPs should be correspond to params.

**Comment by lianxintao** [comment]
Hello, To use auto_pruners_torch.py in NNI, many settings need to pass argparse.ArgumentParser, can you provide setting list for pruners such as AutoCompressPruner, L1FilterPruner, etc., so that the experimental results can be reproduced on NNI. Thank you.

**Comment by suiguoxin** [comment]
> Hello, To use auto_pruners_torch.py in NNI, many settings need to pass argparse.ArgumentParser, can you provide these setting lists for pruners such as AutoCompressPruner, L1FilterPruner, etc., so that the experimental results can be reproduced on NNI. Thank you.

Thanks for your question. To reproduce the results that we present in nni, just use the default config for each pruner. Please refer to [Implementation Details](https://github.com/microsoft/nni/blob/master/docs/en_US/CommunitySharings/ModelCompressionComparison.md#implementation-details). Notice that the performances of your original models (model without pruning) may slightly vary from ours.

**Comment by lianxintao** [comment]
Thank you for your reply, but the results of this experiment are disappointing. I wonder if you have conducted similar experiments on more complex tasks such as target detection, semantic segmentation.Is it possible that the experimental results will be different for more complex tasks.

**Comment by QuanluZhang** [comment]
@lianxintao , we agree that the benchmarking results can be further enriched and improved. We will keep improving it. We highly encourage external contributors to contribute more benchmarking results and good-performing compression algorithms.

**Comment by lianxintao** [comment]
> @lianxintao , we agree that the benchmarking results can be further enriched and improved. We will keep improving it. We highly encourage external contributors to contribute more benchmarking results and good-performing compression algorithms.
Oh, I see, thanks for your reply.
